libp2p/transports/tls/certificate.nim

```diff
@@ -0,0 +1,489 @@
+    addr crt,
+    0, # is_ca
+    -1, # max_pathlen (-1 for no limit)
+  )
+  if ret != 0:
+    raise newException(CertificateCreationError, "Failed to set basic constraints")
+
+  # Set Key Usage
+  ret = mbedtls_x509write_crt_set_key_usage(
+    addr crt, MBEDTLS_X509_KU_DIGITAL_SIGNATURE or MBEDTLS_X509_KU_KEY_ENCIPHERMENT
+  )
+  if ret != 0:
+    raise newException(CertificateCreationError, "Failed to set key usage")
+
+  # Set the MD algorithm
+  mbedtls_x509write_crt_set_md_alg(addr crt, SIGNATURE_ALG)
+
+  # Prepare Buffer for Certificate Serialization
+  const CERT_BUFFER_SIZE = 4096
+  var
+    certBuffer: array[CERT_BUFFER_SIZE, byte]
+    certLen: cint
+
+  # Write the Certificate in DER Format
+  certLen = mbedtls_x509write_crt_der(
+    addr crt,
+    addr certBuffer[0],
+    CERT_BUFFER_SIZE.uint,
+    mbedtls_ctr_drbg_random,
+    addr ctrDrbg,
+  )
+  if certLen < 0:
+    raise newException(
+      CertificateCreationError, "Failed to write certificate in DER format"
+    )
+
+  # Adjust the buffer to contain only the DER data
+  let certificateDer =
+    toSeq(certBuffer[(CERT_BUFFER_SIZE - certLen) ..< CERT_BUFFER_SIZE])
+
+  # Serialize the Private Key in DER Format (PKCS#8)
+  var
+    privKeyBuffer: array[2048, byte]
+    privKeyLen: cint
+
+  privKeyLen = mbedtls_pk_write_key_der(
+    addr certKey, addr privKeyBuffer[0], privKeyBuffer.len.uint
+  )
+  if privKeyLen < 0:
+    raise newException(
+      CertificateCreationError, "Failed to write private key in DER format"
+    )
+
+  # Adjust the buffer to contain only the DER data
+  let privateKeyDer =
+    toSeq(privKeyBuffer[(privKeyBuffer.len - privKeyLen) ..< privKeyBuffer.len])
+
+  # Return the Serialized Certificate and Private Key
+  return (certificateDer, privateKeyDer)
+
+proc libp2pext(
+    p_ctx: pointer,
+    crt: ptr mbedtls_x509_crt,
+    oid: ptr mbedtls_x509_buf,
+    critical: cint,
+    p: ptr byte,
+    endPtr: ptr byte,
+): cint {.cdecl.} =
+  ## Callback function to parse the libp2p extension.
+  ##
+  ## This function is used as a callback by mbedtls during certificate parsing
+  ## to extract the libp2p extension containing the SignedKey.
+  ##
+  ## Parameters:
+  ## - `p_ctx`: Pointer to the P2pExtension object to store the parsed data.
+  ## - `crt`: Pointer to the certificate being parsed.
+  ## - `oid`: Pointer to the OID of the extension.
+  ## - `critical`: Critical flag of the extension.
+  ## - `p`: Pointer to the start of the extension data.
+  ## - `endPtr`: Pointer to the end of the extension data.
+  ##
+  ## Returns:
+  ## - 0 on success, or a negative error code on failure.
+
+  # Check if the OID matches the libp2p extension
+  if oid.len != LIBP2P_EXT_OID_DER.len:
+    return MBEDTLS_ERR_OID_NOT_FOUND # Extension not handled by this callback
+  for i in 0 ..< LIBP2P_EXT_OID_DER.len:
+    if ptrInc(oid.p, i.uint)[] != LIBP2P_EXT_OID_DER[i]:
+      return MBEDTLS_ERR_OID_NOT_FOUND # Extension not handled by this callback
```

Comment:
Maybe use `nimCmpMem`? Not sure if importing just for that is necessary.

Reply:
Where is it defined?